src/main/kotlin/org/elm/ide/intentions/TyFunctionGenerator.kt

```diff
@@ -0,0 +1,117 @@
+import org.elm.lang.core.types.*
+
+
+abstract class TyFunctionGenerator(
+        protected val file: ElmFile,
+        protected val root: Ty
+) {
+    data class GeneratedFunction(
+            val name: String,
+            val paramTy: Ty,
+            val paramName: String,
+            val body: String,
+            val qualifier: String
+    )
+
+    data class Ref(val module: String, val name: String)
+
+    fun TyUnion.toRef() = Ref(module, name)
+    fun AliasInfo.toRef() = Ref(module, name)
+    fun DeclarationInTy.toRef() = Ref(module, name)
+
+    /** All types and aliases referenced in the root ty */
+    protected val declarations by lazy { root.allDeclarations().toList() }
+    /** Additional encoder functions to generate */
+    protected val funcsByTy = mutableMapOf<Ty, GeneratedFunction>()
+    /** Cache of already existing callable expressions that aren't in [funcsByTy] */
+    protected val callablesByTy = mutableMapOf<Ty, String>()
+    /** Unions that need their variants exposed. */
+    protected val unionsToExpose = mutableSetOf<Ref>()
+
+    /** The name of the module in the current file */
+    protected val moduleName by lazy { file.getModuleDecl()?.name ?: "" }
+
+    /** The name to use for the encoder function for each type (does not include the "encoder" prefix) */
+    // There might be multiple types with the same name in different modules, so add the module
+    // name the function for any type with a conflict that isn't defined in this module
+    protected val funcNames by lazy {
+        declarations.groupBy { it.name }
+                .map { (_, decls) ->
+                    decls.map {
+                        it.toRef() to when {
+                            decls.size == 1 -> it.name
+                            else -> it.module.replace(".", "") + it.name
+                        }
+                    }
+                }.flatten().toMap()
+    }
+
+    /** Qualified names of all imported modules */
+    private val importedModules: Set<String> by lazy {
+        file.findChildrenByClass(ElmImportClause::class.java).mapTo(mutableSetOf()) { it.referenceName }
+    }
+
+    /** Get the module qualifier prefix to add to a name */
+    protected fun qualifierFor(ref: Ref): String {
+        return when (ref.module) {
+            moduleName -> ""
+            // We always fully qualify references to modules that we add imports for
+            !in importedModules -> "${ref.module}."
+            else -> ModuleScope.getQualifierForName(file, ref.module, ref.name) ?: ""
+        }
+    }
+
+    /** Prefix [name], defined in [module], with the necessary qualifier */
+    protected fun qual(module: String, name: String) = "${qualifierFor(Ref(module, name))}$name"
+
+    /** Code to insert after the type annotation */
+    abstract val code: String
+
+    /** Imports to add because they are referenced by generated code. */
+    val imports: List<Candidate> by lazy { calculateImports() }
```

Comment:
Lazy properties are admittedly bit of a strange way to structure this. I could instead do something like `abstract fun run(): Pair<String, List<Candidate>>`, although since the generation is stateful, you wouldn't want to call `run` twice. I don't feel strongly either way.

Reply:
Yeah, both sides have their tradeoffs.

I think I would prefer the `run` approach because reading input -> output would be nicer. It seems unlikely that anyone would ever re-run it. And you could add a flag to `TyFunctionGenerator` to keep track of this and throw when someone tries causing trouble.

But if that code comes out a lot worse, than this approach is totally fine with me.